webui/ocw/lib/EC2.py

```diff
@@ -39,27 +27,32 @@ def check_credentials(self):
                 time.sleep(1)
         raise Exception("Invalid EC2 credentials")
 
-    def ec2_resource(self, region='eu-central-1'):
-        if region not in self.__ec2_resource:
-            self.__ec2_resource[region] = boto3.resource('ec2', aws_access_key_id=self.__key,
-                                                         aws_secret_access_key=self.__secret,
+    def get_resource_region(self, region='eu-central-1'):
+        if region not in self.__resource_regions or self.__credentials.isExpired():
+            key, secret = self.getKey_and_secret()
+            self.__resource_regions[region] = boto3.resource('ec2', aws_access_key_id=key,
+                                                             aws_secret_access_key=secret,
+                                                             region_name=region)
+        return self.__resource_regions[region]
+
+    def get_client_region(self, region='eu-central-1'):
+        if region not in self.__client_regions or self.__credentials.isExpired():
+            key, secret = self.getKey_and_secret()
+            self.__client_regions[region] = boto3.client('ec2', aws_access_key_id=key,
+                                                         aws_secret_access_key=secret,
                                                          region_name=region)
-        return self.__ec2_resource[region]
+        return self.__client_regions[region]
 
-    def ec2_client(self, region='eu-central-1'):
-        if region not in self.__ec2_client:
-            self.__ec2_client[region] = boto3.client('ec2', aws_access_key_id=self.__key,
-                                                     aws_secret_access_key=self.__secret,
-                                                     region_name=region)
-        return self.__ec2_client[region]
+    def getKey_and_secret(self):
+        return self.__credentials.getData('access_key'), self.__credentials.getData('secret_key')
 
     def list_instances(self, region='eu-central-1'):
-        return self.ec2_resource(region).instances.all()
+        return self.get_resource_region(region).instances.all()
 
     def list_regions(self):
-        regions_resp = self.ec2_client().describe_regions()
-        regions = [region['RegionName'] for region in regions_resp['Regions']]
-        return regions
+        regions_resp = self.get_client_region().describe_regions()
+        self.__client_regions = [region['RegionName'] for region in regions_resp['Regions']]
```

Comment:
don't know, whay you overwrite the `self.__client_regions` here.
When I understood your change correct, the dict `self__client_regions` is a dict which holds a boto3.client(ec2) object per region.
This `list_regions()` method, is to retrieve all supported regions by given credentials. So a list of strings.